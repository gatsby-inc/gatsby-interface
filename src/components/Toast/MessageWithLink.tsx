/** @jsx jsx */
import * as React from "react"
import { jsx } from "@emotion/core"
import { Link } from "../Link"
import { ThemeCss } from "../../theme"

const messageCss: ThemeCss = theme => ({
  marginBottom: theme.space[3],
})

export interface MessageWithLinkProps {
  linkLabel: React.ReactNode
  href?: string
  target: string
  to?: string
}

export const MessageWithLink: React.FC<MessageWithLinkProps> = ({
  children,
  linkLabel,
  ...linkProps
}) => (
  <React.Fragment>
    <div css={messageCss}>{children}</div>
    <Link
      // we need to cast to "any" because TS claims that "href" is required for <Link /> even though it's not
      {...(linkProps as any)}
    >
      {linkLabel}
    </Link>
  </React.Fragment>
)
