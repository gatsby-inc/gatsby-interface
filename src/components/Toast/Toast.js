/** @jsx jsx */
import { jsx } from "@emotion/core"
import PropTypes from "prop-types"
import Alert from "@reach/alert"
import { keyframes, css } from "@emotion/core"
import { MdDone, MdClose, MdWarning } from "react-icons/md"
import {
  fontSizes,
  dimensions,
  palette,
  radius,
  spaces,
} from "../../utils/presets"
import { ToastTones } from "./constants"

const toastEntryAnimation = keyframes`
  100% {
     transform: perspective(1000px) rotateX(0);
  }
`

const toastCss = css`
  align-items: center;
  animation: ${toastEntryAnimation} 0.5s 0.25s ease forwards;
  background: ${palette.grey[900]};
  border-left: 8px solid ${palette.green[500]};
  border-radius: ${radius.default} ${radius.default} 0 0;
  color: ${palette.green[50]};
  display: flex;
  font-size: ${fontSizes.xs};
  min-height: ${dimensions.toast.minHeight};
  max-width: calc(100% - (${spaces.l} * 2));
  padding-left: ${spaces.s};
  transform: perspective(1000px) rotateX(90deg);
  transform-origin: bottom center;

  svg {
    height: auto;
    width: calc(${dimensions.toast.minHeight} * 0.4);
  }

  &:not(:first-of-type) {
    border-radius: ${radius.default};
    margin-bottom: ${spaces[`3xs`]};
  }
`

const messageCss = css`
  line-height: 1;
  margin: 0 ${spaces[`2xs`]} 0 ${spaces.xs};
`

const statusCss = css`
  align-items: center;
  color: ${palette.green[500]};
  display: flex;
`

const closeButtonCss = css`
  align-items: center;
  background: none;
  border: none;
  color: ${palette.grey[400]};
  cursor: pointer;
  display: flex;
  height: ${dimensions.toast.minHeight};
  justify-content: center;
  width: ${dimensions.toast.minHeight};
`

const toastColorByTone = {
  SUCCESS: palette.green[500],
  DANGER: palette.red[600],
}

const ToastIconByTone = {
  SUCCESS: MdDone,
  DANGER: MdWarning,
}

export default function Toast({
  id,
  message,
  tone,
  onRemove,
  closeButtonLabel,
}) {
  const IconComponent = ToastIconByTone[tone]

  return (
    <Alert
      css={[
        toastCss,
        css({
          borderLeftColor: toastColorByTone[tone],
        }),
      ]}
      data-testid="toast"
      type={tone === `DANGER` ? `assertive` : `polite`}
    >
      <span
        css={[
          statusCss,
          css({
            color: toastColorByTone[tone],
          }),
        ]}
      >
        <IconComponent />
      </span>
      <div css={messageCss}>{message}</div>
      <button
        css={closeButtonCss}
        type="button"
        onClick={() => onRemove(id)}
        aria-label={closeButtonLabel}
      >
        <MdClose />
      </button>
    </Alert>
  )
}

Toast.propTypes = {
  id: PropTypes.number.isRequired,
  message: PropTypes.node.isRequired,
  tone: PropTypes.oneOf(ToastTones),
  onRemove: PropTypes.func.isRequired,
  closeButtonLabel: PropTypes.string.isRequired,
}
