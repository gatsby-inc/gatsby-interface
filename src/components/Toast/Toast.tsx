/** @jsx jsx */
import * as React from "react"
import Alert from "@reach/alert"
import { jsx } from "@emotion/core"
import { MdDone, MdClose } from "react-icons/md"

import { ToastTone } from "./types"
import { Theme, ThemeCss } from "../../theme"
import { ThemeProvider } from "../ThemeProvider"

const MIN_HEIGHT = `3rem`

const toastCss: ThemeCss = theme => ({
  display: "flex",
  alignItems: "center",
  background: theme.colors.grey[90],
  borderLeft: `${theme.space[3]} solid ${theme.colors.green[50]}`,
  borderRadius: theme.radii[2],
  color: theme.colors.green[5],
  fontSize: theme.fontSizes[1],
  minHeight: MIN_HEIGHT,
  paddingLeft: theme.space[4],
})

const iconCss: ThemeCss = theme => ({
  fontSize: theme.fontSizes[4],
})

const messageCss: ThemeCss = theme => ({
  lineHeight: theme.lineHeights.solid,
  marginTop: 0,
  marginBottom: 0,
  marginRight: theme.space[2],
  marginLeft: theme.space[3],
})

const statusCss: ThemeCss = theme => ({
  display: "flex",
  alignItems: "center",
  color: theme.colors.green[50],
})

const closeButtonCss: ThemeCss = theme => ({
  display: "flex",
  alignItems: "center",
  background: "none",
  border: "none",
  color: theme.colors.grey[40],
  cursor: "pointer",
  height: MIN_HEIGHT,
  justifyContent: "center",
  width: MIN_HEIGHT,
  marginLeft: `auto`,
})

function getToastColorByTone(theme: Theme, tone: ToastTone): string {
  if (tone === `SUCCESS`) {
    return theme.colors.green[50]
  }
  throw new Error(`Unknown "tone" for Toast component: ${tone}`)
}

const ToastIconByTone: Record<ToastTone, React.ComponentType> = {
  SUCCESS: MdDone,
}

export interface ToastProps {
  message: React.ReactNode
  onClose: () => void
  closeButtonLabel: string
  tone: ToastTone
  className?: string
  style?: React.CSSProperties
}

export function Toast({
  message,
  tone,
  closeButtonLabel,
  onClose,
  className,
  style,
}: ToastProps) {
  const IconComponent = ToastIconByTone[tone]

  const finalToastCss: ThemeCss = theme => [
    toastCss(theme),
    {
      borderLeftColor: getToastColorByTone(theme, tone),
    },
  ]

  const finalStatusCss: ThemeCss = theme => [
    statusCss(theme),
    {
      color: getToastColorByTone(theme, tone),
    },
  ]

  return (
    <Alert
      css={finalToastCss}
      data-testid="toast"
      type="polite"
      className={className}
      style={style}
    >
      {/**
       * We have to wrap Alert content in ThemeProvider, since ReachUI's Alert does not allow to use context that is defined outside of Alert
       *
       * See this issue: https://github.com/reach/reach-ui/issues/540
       * Source code: https://github.com/reach/reach-ui/blob/develop/packages/alert/src/index.tsx
       * */}
      <ThemeProvider>
        <span css={finalStatusCss}>
          <IconComponent css={iconCss} />
        </span>
        <div css={messageCss}>{message}</div>
        <button
          css={closeButtonCss}
          type="button"
          onClick={onClose}
          aria-label={closeButtonLabel}
        >
          <MdClose css={iconCss} />
        </button>
      </ThemeProvider>
    </Alert>
  )
}
