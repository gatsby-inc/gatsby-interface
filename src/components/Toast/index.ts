export { ToastProvider, ToastConsumer } from "./ToastProvider"
export { useShowSuccessToast, useShowToast, useToastActions } from "./hooks"
