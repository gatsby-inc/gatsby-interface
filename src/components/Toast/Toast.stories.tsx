import * as React from "react"
import { Meta, Story } from "@storybook/react"
import { Button } from "../Button"
import { ToastProvider, ToastConsumer, useShowSuccessToast } from "."
import { Toast, ToastProps } from "./Toast"

export default {
  title: `Toast`,
  component: ToastProvider,
  subcomponents: {
    ToastConsumer,
  },
  parameters: {
    componentSubtitle:
      "Toasts provide brief messages about app processes at the bottom of the screen, usually to give feedback after an action has taken place.",
  },
} as Meta

const Template: Story<ToastProps> = args => <Toast {...args} />

export const Basic = Template.bind({})

Basic.args = {
  message: `Some short toast message`,
  closeButtonLabel: `Close`,
  onClose: () => undefined,
  tone: `SUCCESS`,
}

export const TriggerToast = () => {
  function NoTimeoutExample() {
    const showSuccessToast = useShowSuccessToast()

    return (
      <Button
        onClick={() =>
          showSuccessToast(`This message will stay on screen until closed`, {
            timeout: 0,
          })
        }
      >
        Show toast without auto hide
      </Button>
    )
  }
  return (
    <ToastProvider>
      <ToastConsumer>
        {({ showToast }) => (
          <React.Fragment>
            <Button onClick={() => showToast(`Your action was successful`)}>
              Show toast
            </Button>
          </React.Fragment>
        )}
      </ToastConsumer>
      <NoTimeoutExample />
    </ToastProvider>
  )
}
