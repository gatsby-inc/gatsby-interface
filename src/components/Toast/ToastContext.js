/** @jsx jsx */
import { jsx } from "@emotion/core"
import React from "react"
import PropTypes from "prop-types"
import { css } from "@emotion/core"
import Toast from "./Toast"
import { zIndices } from "../../utils/presets"

export const ToastContext = React.createContext()
export const ToastConsumer = ToastContext.Consumer

const containerCss = css`
  align-items: center;
  bottom: 0;
  display: flex;
  flex-direction: column-reverse;
  left: 50%;
  position: fixed;
  transform: translate(-50%, 0);
  width: 100%;
  z-index: ${zIndices.toasts};
`

const DEFAULT_TIMEOUT = 5000
const DEFAULT_TONE = `SUCCESS`

export function ToastProvider({ children, closeButtonLabel = `Close` }) {
  const [toasts, setToasts] = React.useState([])

  const timeoutsRef = React.useRef({})

  const removeToast = React.useCallback(toastId => {
    setToasts(prevToasts => prevToasts.filter(({ id }) => id !== toastId))
    clearTimeout(timeoutsRef.current[toastId])
    delete timeoutsRef.current[toastId]
  }, [])

  const showToast = React.useCallback(
    (message, { tone = DEFAULT_TONE, timeout = DEFAULT_TIMEOUT } = {}) => {
      const toastId = Math.random()
      setToasts(prevToasts => [...prevToasts, { id: toastId, message, tone }])

      if (timeout > 0) {
        timeoutsRef.current[toastId] = setTimeout(() => {
          removeToast(toastId)
        }, timeout)
      }
    },
    []
  )

  const contextValue = React.useMemo(() => {
    return {
      showToast,
    }
  }, [showToast])

  return (
    <ToastContext.Provider value={contextValue}>
      {children}
      <div css={containerCss}>
        {toasts.map(toast => (
          <Toast
            key={toast.id}
            {...toast}
            onRemove={removeToast}
            closeButtonLabel={closeButtonLabel}
          />
        ))}
      </div>
    </ToastContext.Provider>
  )
}

ToastProvider.propTypes = {
  children: PropTypes.node.isRequired,
  closeButtonLabel: PropTypes.string,
}

export function useToastContext() {
  const context = React.useContext(ToastContext)

  if (!context) {
    throw new Error(
      `Toast hooks cannot be used outside the ToastProvider component`
    )
  }

  return context
}
