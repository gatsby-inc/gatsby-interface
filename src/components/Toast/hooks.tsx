import * as React from "react"
import { ToastOptions, useToastContext } from "./ToastContext"
import { DEFAULT_TIMEOUT, DEFAULT_TONE } from "./constants"
import { ToastTone } from "./types"

export const useShowToast = () => {
  const { showToast } = useToastContext()

  return showToast
}

export const useShowSuccessToast = () => {
  const showToast = useShowToast()

  return React.useCallback(
    (
      message: React.ReactNode,
      options: Omit<Partial<ToastOptions>, "tone"> = {}
    ) => {
      showToast(message, { ...options, tone: `SUCCESS` })
    },
    [showToast]
  )
}

export interface ToastData {
  id: symbol
  message: React.ReactNode
  tone: ToastTone
}

export const useToastActions = () => {
  const [toasts, setToasts] = React.useState<ToastData[]>([])
  const timeoutsRef = React.useRef<Map<symbol, number>>(new Map())

  const removeToast = React.useCallback((toastId: symbol) => {
    setToasts(prevToasts => prevToasts.filter(({ id }) => id !== toastId))

    window.clearTimeout(timeoutsRef.current.get(toastId))

    timeoutsRef.current.delete(toastId)
  }, [])

  const showToast = React.useCallback(
    (
      message: React.ReactNode,
      {
        tone = DEFAULT_TONE,
        timeout = DEFAULT_TIMEOUT,
      }: Partial<ToastOptions> = {}
    ) => {
      const toastId = Symbol(`toast`)

      setToasts(prevToasts => [...prevToasts, { id: toastId, message, tone }])

      if (timeout > 0) {
        const timeOutId = window.setTimeout(() => {
          removeToast(toastId)
        }, timeout)

        timeoutsRef.current.set(toastId, timeOutId)
      }
    },
    []
  )

  React.useEffect(() => {
    // Clear all timeouts
    return () => {
      for (const timeout of timeoutsRef.current.values()) {
        window.clearTimeout(timeout)
      }
    }
  }, [])

  return { toasts, showToast, removeToast }
}
