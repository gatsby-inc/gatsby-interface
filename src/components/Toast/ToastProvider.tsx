/** @jsx jsx */
import { jsx } from "@emotion/core"
import * as React from "react"
import { animated, useSpring } from "react-spring"
import { Toast, ToastProps } from "./Toast"
import { ToastContext } from "./ToastContext"
import { useToastActions } from "./hooks"
import { ThemeCss } from "../../theme"

export const ToastConsumer = ToastContext.Consumer

const containerCss: ThemeCss = theme => ({
  alignItems: "center",
  bottom: 0,
  display: "flex",
  flexDirection: "column-reverse",
  left: "50%",
  position: "fixed",
  transform: "translate(-50%, 0)",
  width: "100%",
  zIndex: theme.zIndices.base,
  pointerEvents: `none`,
  paddingRight: theme.space[7],
  paddingLeft: theme.space[7],
})

const toastCss: ThemeCss = theme => ({
  pointerEvents: `auto`,
  marginBottom: theme.space[1],

  "&:first-of-type": {
    borderBottomLeftRadius: 0,
    borderBottomRightRadius: 0,
    marginBottom: 0,
  },
})

const AnimatedToast = animated(Toast)

function ToastItem(props: ToastProps) {
  const styleProps = useSpring({
    from: {
      transform: `perspective(1000px) rotateX(90deg)`,
      transformOrigin: "bottom center",
    },
    to: {
      transform: `perspective(1000px) rotateX(0)`,
      transformOrigin: "bottom center",
    },
  })

  return <AnimatedToast style={styleProps} {...props} />
}

export interface ToastProviderProps {
  children: React.ReactNode
  closeButtonLabel?: string
}

export function ToastProvider({
  children,
  closeButtonLabel = `Close`,
}: ToastProviderProps) {
  const { toasts, showToast, removeToast } = useToastActions()

  const contextValue = React.useMemo(() => {
    return {
      showToast,
    }
  }, [showToast])

  return (
    <ToastContext.Provider value={contextValue}>
      {children}

      <div css={containerCss}>
        {toasts.map((toast, index) => (
          <ToastItem
            key={`toast-${index}`} // this is probably not the best, but we can't use symbol as key :(
            {...toast}
            onClose={() => removeToast(toast.id)}
            closeButtonLabel={closeButtonLabel}
            css={toastCss}
          />
        ))}
      </div>
    </ToastContext.Provider>
  )
}
