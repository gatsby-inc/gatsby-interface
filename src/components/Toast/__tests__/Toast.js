import * as React from "react"
import { fireEvent, act } from "@testing-library/react"
import { ToastProvider, ToastConsumer, useShowSuccessToast } from "../"
import { Toast } from "../Toast"
import { renderWithTheme } from "../../../utils/testing"

const render = renderWithTheme

describe(`Toast`, () => {
  const baseProps = {
    message: `Lorem ipsum`,
    onClose: () => {
      // do nothing.
    },
    closeButtonLabel: "close",
    tone: `SUCCESS`,
    id: 0,
  }

  it(`displays the toast message`, async () => {
    const { queryByText } = render(<Toast {...baseProps} />)

    expect(queryByText(`Lorem ipsum`)).toBeTruthy()
  })

  it(`has accessible label for the close button`, async () => {
    const { container } = render(<Toast {...baseProps} />)

    expect(container.querySelector(`button`)).toHaveAttribute(
      `aria-label`,
      `close`
    )
  })

  it(`calls onRemove callback when clicking the close button`, async () => {
    const removeFn = jest.fn()
    const { getByLabelText } = render(
      <Toast {...baseProps} onClose={removeFn} />
    )

    fireEvent.click(getByLabelText(`close`))

    expect(removeFn).toHaveBeenCalled()
  })
})

describe(`ToastProvider`, () => {
  it(`exposes toast tools via context`, () => {
    const renderFn = jest.fn()

    render(
      <ToastProvider>
        <ToastConsumer>{renderFn}</ToastConsumer>
      </ToastProvider>
    )

    expect(renderFn).toHaveBeenCalledWith(
      expect.objectContaining({
        showToast: expect.any(Function),
      })
    )
  })

  function TestComponent({ toastOptions = {} }) {
    return (
      <ToastConsumer>
        {({ showToast }) => (
          <button onClick={() => showToast(`Lorem ipsum`, toastOptions)}>
            Show toast
          </button>
        )}
      </ToastConsumer>
    )
  }

  it(`allows to specify custom label for the toast close button`, () => {
    const { getByText, getByTestId } = render(
      <ToastProvider closeButtonLabel="Zamknąć">
        <TestComponent toastOptions={{ timeout: 0 }} />
      </ToastProvider>
    )

    act(() => {
      fireEvent.click(getByText(`Show toast`))
    })

    expect(getByTestId(`toast`).querySelector(`button`)).toHaveAttribute(
      `aria-label`,
      `Zamknąć`
    )
  })

  it(`removes a toast on timeout`, () => {
    const { getByText, queryByTestId } = render(
      <ToastProvider>
        <TestComponent toastOptions={{ timeout: 500 }} />
      </ToastProvider>
    )

    act(() => {
      fireEvent.click(getByText(`Show toast`))
    })

    expect(queryByTestId(`toast`)).toBeTruthy()

    setTimeout(() => expect(queryByTestId(`toast`)).toBeFalsy(), 500)
  })

  it(`persists a toast if the provided timeout is 0`, () => {
    const { getByText, queryByTestId } = render(
      <ToastProvider>
        <TestComponent toastOptions={{ timeout: 0 }} />
      </ToastProvider>
    )

    act(() => {
      fireEvent.click(getByText(`Show toast`))
    })

    expect(queryByTestId(`toast`)).toBeTruthy()

    setTimeout(() => expect(queryByTestId(`toast`)).toBeTruthy(), 600)
  })
})

describe(`useShowSuccessToast hook`, () => {
  it(`returns a method to show success toast`, () => {
    function TestComponent() {
      const showSuccessToast = useShowSuccessToast()

      return (
        <button onClick={() => showSuccessToast(`Lorem ipsum`)}>
          Show toast
        </button>
      )
    }

    const { getByText, queryByTestId } = render(
      <ToastProvider>
        <TestComponent />
      </ToastProvider>
    )

    act(() => {
      fireEvent.click(getByText(`Show toast`))
    })

    expect(queryByTestId(`toast`)).toBeTruthy()
  })
})
