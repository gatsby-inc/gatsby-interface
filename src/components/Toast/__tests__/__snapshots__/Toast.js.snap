// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders unchanged 1`] = `
@keyframes animation-0 {
  100% {
    -webkit-transform: perspective(1000px) rotateX(0);
    -ms-transform: perspective(1000px) rotateX(0);
    transform: perspective(1000px) rotateX(0);
  }
}

@keyframes animation-0 {
  100% {
    -webkit-transform: perspective(1000px) rotateX(0);
    -ms-transform: perspective(1000px) rotateX(0);
    transform: perspective(1000px) rotateX(0);
  }
}

.emotion-3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-animation: animation-0 0.5s 0.25s ease forwards;
  animation: animation-0 0.5s 0.25s ease forwards;
  background: #232129;
  border-left: 8px solid #37b635;
  border-radius: 4px 4px 0 0;
  color: #f7fdf7;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  min-height: 3rem;
  max-width: calc(100% - (1.5rem * 2));
  padding-left: 0.75rem;
  -webkit-transform: perspective(1000px) rotateX(90deg);
  -ms-transform: perspective(1000px) rotateX(90deg);
  transform: perspective(1000px) rotateX(90deg);
  -webkit-transform-origin: bottom center;
  -ms-transform-origin: bottom center;
  transform-origin: bottom center;
  border-left-color: #37b635;
}

.emotion-3 svg {
  height: auto;
  width: calc(3rem * 0.4);
}

.emotion-3:not(:first-of-type) {
  border-radius: 4px;
  margin-bottom: 0.125rem;
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #37b635;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #37b635;
}

.emotion-1 {
  line-height: 1;
  margin: 0 0.25rem 0 0.5rem;
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: none;
  border: none;
  color: #b7b5bd;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 3rem;
}

.emotion-7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-animation: animation-0 0.5s 0.25s ease forwards;
  animation: animation-0 0.5s 0.25s ease forwards;
  background: #232129;
  border-left: 8px solid #37b635;
  border-radius: 4px 4px 0 0;
  color: #f7fdf7;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  min-height: 3rem;
  max-width: calc(100% - (1.5rem * 2));
  padding-left: 0.75rem;
  -webkit-transform: perspective(1000px) rotateX(90deg);
  -ms-transform: perspective(1000px) rotateX(90deg);
  transform: perspective(1000px) rotateX(90deg);
  -webkit-transform-origin: bottom center;
  -ms-transform-origin: bottom center;
  transform-origin: bottom center;
  border-left-color: #ec1818;
}

.emotion-7 svg {
  height: auto;
  width: calc(3rem * 0.4);
}

.emotion-7:not(:first-of-type) {
  border-radius: 4px;
  margin-bottom: 0.125rem;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #37b635;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #ec1818;
}

<div>
  <div
    class="emotion-3"
    data-reach-alert="true"
    data-testid="toast"
  >
    <span
      class="emotion-0"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
        />
      </svg>
    </span>
    <div
      class="emotion-1"
    >
      Lorem ipsum
    </div>
    <button
      aria-label="close"
      class="emotion-2"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
  </div>
  <div
    class="emotion-7"
    data-reach-alert="true"
    data-testid="toast"
  >
    <span
      class="emotion-4"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        />
      </svg>
    </span>
    <div
      class="emotion-1"
    >
      Lorem ipsum
    </div>
    <button
      aria-label="close"
      class="emotion-2"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
  </div>
</div>
`;
